### PR TITLE
EPMRPP-59875 Add a new button to History Line

### PR DESCRIPTION
**Pull request by katrin-kot**

### CODE REVIEW

File: app/src/pages/inside/logsPage/historyLine/historyLine.jsx [hunk 1]
```diff
@@ -66,6 +78,29 @@ export class HistoryLine extends Component {
       <div className={cx('history-line')}>
         <ScrollWrapper autoHeight hideTracksWhenNotNeeded>
           <div className={cx('history-line-items')}>
+            {historyItems.length === DEFAULT_HISTORY_DEPTH && (
+              <GhostButton
+                style={{
```
Comment (ElenaRomanchuk) [inline app/src/pages/inside/logsPage/historyLine/historyLine.jsx:83]: Most likely you need to reuse one of the existing button components. Overwise you should create the new one. I suggest to research if the needed component exists, overwise discuss with the designer are you really need a new one.
Never use inline styles unnecessarily.

File: app/src/pages/inside/logsPage/historyLine/historyLine.jsx [hunk 2]
```diff
@@ -66,6 +78,29 @@ export class HistoryLine extends Component {
...
+                  borderRadius: '4px',
+                  boxShadow: '0 2px 6px 0 rgba(0, 0, 0, 0.07)',
+                }}
+              >
+                {this.state.isLoading ? (
+                  <div className={cx('spinner-wrapper')}>
+                    <SpinningPreloader />
+                  </div>
+                ) : (
+                  HISTORY_LINE_LOAD_MORE_BTN_TITLE
```
Comment (ElenaRomanchuk) [inline app/src/pages/inside/logsPage/historyLine/historyLine.jsx:100]: Use react-intl to localize messages and labels. Never hardcode them as a constant unnecessarily.

**Review by ElenaRomanchuk — CHANGES_REQUESTED** [review]

File: app/src/controllers/log/sagas.js [hunk 3]
```diff
@@ -73,6 +76,20 @@ function* fetchActivity() {
   yield take(createFetchPredicate(ACTIVITY_NAMESPACE));
 }
 
+function* fetchMoreHistoryItems() {
```
Comment (AmsterGet) [inline app/src/controllers/log/sagas.js:79]: Seems that this saga is too similar with `fetchHistoryItems` [saga](https://github.com/reportportal/service-ui/pull/2601/files#diff-f331d34263a8765d03a750f31a61d1d83cd24a2109cb6164d648eb1e72f62c52L111).
Can we reuse `fetchHistoryItems` saga to handle action with additional payload parameter (f.e. `loadMore: true`)?

File: app/src/pages/inside/logsPage/historyLine/historyLine.jsx [hunk 4]
```diff
@@ -39,33 +55,68 @@ const cx = classNames.bind(styles);
...
+  loadMoreItems = () => {
+    this.setState({ isLoading: true });
+    this.props.fetchMoreItemsHistoryAction();
+    this.setState({ isLoading: false });
+  };
+
   render() {
-    const { historyItems, activeItemId, changeActiveItem } = this.props;
+    const { historyItems, activeItemId, changeActiveItem, intl } = this.props;
+    const shouldShowLoadMore =
```
Comment (AmsterGet) [inline app/src/pages/inside/logsPage/historyLine/historyLine.jsx:99]: Will these conditions handle the case where the maximum allowed history depth is 30?

**Review by AmsterGet — CHANGES_REQUESTED** [review]

File: app/src/pages/inside/logsPage/historyLine/historyLine.jsx [hunk 5]
```diff
@@ -39,33 +55,67 @@ const cx = classNames.bind(styles);
...
   };
 
   checkIfTheItemLinkIsActive = (item) =>
     item.id !== this.props.activeItemId && item.status !== NOT_FOUND;
 
-  render() {
-    const { historyItems, activeItemId, changeActiveItem } = this.props;
+  loadMoreItems = () => {
+    this.setState({ isLoading: true });
+    this.props.fetchMoreItemsHistoryAction(true);
```
Comment (AmsterGet) [inline app/src/pages/inside/logsPage/historyLine/historyLine.jsx:93]: Seems that this method always should fetch **more** history items, so please avoid misunderstandings and use one clear variant `fetchHistoryItemsAction({ loadMore: true })` or `fetchMoreHistoryItemsAction()`.

File: app/src/pages/inside/logsPage/historyLine/historyLine.scss [hunk 6]
```diff
@@ -30,3 +30,45 @@
...
+.load-more {
+  position: relative;
+  display: inline-flex;
+  justify-content: center;
+  width: 76px;
+  height: 57px;
+  padding: 11px 21px 10px;
+  margin-right: 20px;
+  font-size: 13px;
+  font-weight: 600;
```
Comment (AmsterGet) [inline app/src/pages/inside/logsPage/historyLine/historyLine.scss:47]: Please remove `font-weight` and use `$FONT-SEMIBOLD`

File: app/src/pages/inside/logsPage/historyLine/historyLine.scss [hunk 7]
```diff
@@ -30,3 +30,45 @@
...
+  vertical-align: top;
+  background-color: $COLOR--white-two;
+  cursor: pointer;
+  &:hover {
+    background-color: lighten($COLOR--topaz, 10%);
+    border-color: lighten($COLOR--topaz, 10%);
+    color: $COLOR--white-two;
+  }
+
+  &.active,
```
Comment (AmsterGet) [inline app/src/pages/inside/logsPage/historyLine/historyLine.scss:68]: According to the component you don't use `.active` class name nowhere, so please remove this line.

File: app/src/controllers/log/sagas.js [hunk 8]
```diff
@@ -108,14 +111,19 @@ function* fetchStackTrace({ payload: logItem }) {
...
 }
 
-function* fetchHistoryItems() {
+function* fetchHistoryItems({ payload } = { payload: false }) {
   const activeProject = yield select(activeProjectSelector);
   const logItemId = yield select(logItemIdSelector);
+  const countOfHistoryItems = yield select(historyItemsSelector);
   const isAllLaunches = yield select(includeAllLaunchesSelector);
   const historyLineMode = isAllLaunches ? HISTORY_LINE_TABLE_MODE : HISTORY_LINE_DEFAULT_VALUE;
+  let historyDepth;
```
Comment (AmsterGet) [inline app/src/controllers/log/sagas.js:120]: Why not just:
```
const historyDepth = payload
    ? countOfHistoryItems.length + NUMBER_OF_ITEMS_TO_LOAD
    : DEFAULT_HISTORY_DEPTH;
```

File: app/src/controllers/log/sagas.js [hunk 9]
```diff
@@ -108,14 +111,19 @@ function* fetchStackTrace({ payload: logItem }) {
   yield take(createFetchPredicate(STACK_TRACE_NAMESPACE));
 }
 
-function* fetchHistoryItems() {
+function* fetchHistoryItems({ payload } = { payload: false }) {
   const activeProject = yield select(activeProjectSelector);
   const logItemId = yield select(logItemIdSelector);
+  const countOfHistoryItems = yield select(historyItemsSelector);
```
Comment (AmsterGet) [inline app/src/controllers/log/sagas.js:117]: For now it is the `historyItems`, not the count of them.

File: app/src/pages/inside/logsPage/historyLine/historyLine.jsx [hunk 10]
```diff
@@ -39,33 +55,67 @@ const cx = classNames.bind(styles);
...
+    this.setState({ isLoading: false });
```
Comment (AmsterGet) [inline app/src/pages/inside/logsPage/historyLine/historyLine.jsx:94]: It looks like this `loading` will be stopped before the response is received from BE.
For the fast fix I suggest to use a `callback` in the action to set `isLoading: false`, when the response will be received.

**Comment by AmsterGet** [comment]
Closed in favor of https://github.com/reportportal/service-ui/pull/2606

**Pull request closed without merging**